In [2]:
#from langchain.output_parsers import StructuredOutputParser
from langchain.output_parsers import ResponseSchema
from langchain.prompts import ChatPromptTemplate
from langchain.chat_models import ChatOpenAI
import datetime
import os
import openai

from dotenv import load_dotenv, find_dotenv
_ = load_dotenv(find_dotenv())
openai.api_key = os.environ['OPENAI_API_KEY']

In [3]:
llm_model = "gpt-4"


def get_completion(prompt, model=llm_model):
    messages = [{"role": "user", "content": prompt}]
    response = openai.ChatCompletion.create(
        model=model,
        messages=messages,
        temperature=0,
    )
    return response.choices[0].message["content"]

In [4]:
from langchain.document_loaders import PyPDFLoader

loader = PyPDFLoader("docs/Biometric Vehicle Access System.pdf")
pdf_abstract = loader.load_and_split()

In [5]:
abstract1 = """
The Biometric Vehicle Access System (BVAS) is an innovative technology designed to revolutionize traditional vehicle security and access methods. This system employs advanced biometric authentication, including fingerprint and facial recognition, to ensure secure and convenient entry and ignition processes. BVAS enhances vehicle security by replacing traditional key-based and electronic fob systems with a seamless and personalized biometric verification process. The technology integrates biometric sensors into door handles, steering wheels, and ignition systems, allowing for quick and reliable user authentication. The BVAS not only provides an additional layer of security against unauthorized access but also enhances user convenience by eliminating the need for physical keys or key fobs. Users can effortlessly unlock, start, and operate their vehicles through a simple and rapid biometric scan. The system is designed with robust anti-spoofing measures to prevent unauthorized access attempts. Furthermore, BVAS contributes to the growing trend of biometric integration in smart vehicles, aligning with the industry's commitment to innovation, user experience, and safety. As vehicles continue to evolve into interconnected and autonomous entities, BVAS sets a new standard for personalized and secure access, catering to the increasing demand for sophisticated yet user-friendly solutions in the automotive sector.
"""

In [6]:
abstract2 = """
The Biometric Vehicle Access System (BVAS) employs advanced fingerprint and facial recognition for secure and convenient vehicle entry. It replaces traditional key and fob systems with seamless biometric verification integrated into handles, steering wheels, and ignition. BVAS enhances security, eliminates the need for physical keys, and allows users to unlock and start their vehicles with a rapid biometric scan. The system includes anti-spoofing measures and aligns with the growing trend of biometric integration in smart vehicles, setting a new standard for personalized and secure access in the automotive sector.
"""

In [7]:
abstract3 = """ 
BVAS revolutionizes vehicle security with advanced biometric authentication, eliminating keys for convenient entry. Integrated into handles and steering wheels, it sets a new standard for secure, user-friendly access, aligning with the trend of biometric integration in smart vehicles. 
"""

In [8]:
prompt1 = f"""```{pdf_abstract}```\
The abstract above describes a concept for a novel invention.\
I would like to search a patent database to find out whether \
there are already patents for such a concept. Name 5 phrases that I can \
use for the search. Each phrase should contain between 5 to 10 words. \
Optimize the phrases to get back more results.
"""

In [9]:
prompt2 = f"""```{pdf_abstract}```\
The abstract above describes a concept for a novel invention.\
I would like to search a patent database to find out whether \
there are already patents for such a concept. Please list me the codes of the 5 most relevant \
USPTO classifications to a possible patent for this concept without explanations for the codes.
"""

In [10]:
response_keywords = get_completion(prompt1)

In [11]:
response_classes = get_completion(prompt2)

In [12]:
print(response_keywords)
print("----")
print(response_classes)

1. "Biometric Vehicle Access System technology"
2. "Advanced biometric authentication in vehicles"
3. "Biometric sensors in door handles and ignition systems"
4. "Anti-spoofing measures in vehicle access"
5. "Biometric integration in smart vehicles"
----
1. G06K9/00
2. B60R25/10
3. G07C9/00
4. E05B49/00
5. G06F21/32


In [13]:
import json

f = open('data_dump/test.json')

data = json.load(f)

In [14]:
for i in data['organic_results']:

    print("Result", i['position'], ": ", i['title'])
    # print(i['snippet'], "\n")
    print(i['pdf'], "\n")

Result 1 :  Method and system for localizing parts of an object in an image for computer …
https://patentimages.storage.googleapis.com/a2/95/38/bcc3b288e56669/US9275273.pdf 

Result 2 :  Adaptive multi-modal integrated biometric identification and surveillance …
https://patentimages.storage.googleapis.com/98/f0/22/ae169ef67be168/US9432632.pdf 

Result 3 :  Spoof detection for biometric authentication
https://patentimages.storage.googleapis.com/84/29/92/5fe2a153298d2d/US9971920.pdf 

Result 4 :  Liveness testing methods and apparatuses and image processing methods and …
https://patentimages.storage.googleapis.com/b1/98/34/94c48fd1f99eeb/US11151397.pdf 

Result 5 :  Biometrics based on locally consistent features
https://patentimages.storage.googleapis.com/a6/11/c7/5583de7c8fb29d/US9060688.pdf 

Result 6 :  System and method for detecting the authenticity of products
https://patentimages.storage.googleapis.com/9c/e5/fc/eb90d460e518de/US10956732.pdf 

Result 7 :  System and process for au

In [15]:
pdf_list = []
for i in data['organic_results']:
    loader = PyPDFLoader(i['pdf'])
    pdf = loader.load_and_split()
    pdf_list.append(pdf[0])

    

In [16]:
test_py = pdf_list[1]

print(test_py)

page_content='(12) United States Patent \n Cheng et al. USOO9432632B2 \n US 9.432,632 B2 \n Aug. 30, 2016 (10) Patent No.: \n (45) Date of Patent: \n (54) ADAPTIVE MULTI-MODAL INTEGRATED \n BOMETRIC IDENTIFICATION AND \n SURVELLANCE SYSTEMS \n (71) Applicant: Proximex Corporation, Cupertino, CA \n (US) \n (72) Inventors: Ken P. Cheng, Saratoga, CA (US); \n Edward Y. Chang, Santa Barbara, CA (US); Yuan-Fang Wang, Goleta, CA \n (US) \n (73) Assignee: Proximex Corporation, Cupertino, CA \n (US) \n Subject to any disclaimer, the term of this patent is extended or adjusted under 35 \n U.S.C. 154(b) by 0 days. \n (21) Appl. No.: 14/607,201 \n (22) Filed: Jan. 28, 2015 \n (65) Prior Publication Data \n US 2015/O138332 A1 May 21, 2015 \n Related U.S. Application Data \n (60) Division of application No. 13/738,655, filed on Jan. \n 10, 2013, now Pat. No. 8,976,237, and a continuation of application No. 13/101,149, filed on May 5, 2011, \n now Pat. No. 8.373,753, and a division of application (*

page_content='(12) United States Patent \n Cheng et al. USOO9432632B2 \n US 9.432,632 B2 \n Aug. 30, 2016 (10) Patent No.: \n (45) Date of Patent: \n (54) ADAPTIVE MULTI-MODAL INTEGRATED \n BOMETRIC IDENTIFICATION AND \n SURVELLANCE SYSTEMS \n (71) Applicant: Proximex Corporation, Cupertino, CA \n (US) \n (72) Inventors: Ken P. Cheng, Saratoga, CA (US); \n Edward Y. Chang, Santa Barbara, CA (US); Yuan-Fang Wang, Goleta, CA \n (US) \n (73) Assignee: Proximex Corporation, Cupertino, CA \n (US) \n Subject to any disclaimer, the term of this patent is extended or adjusted under 35 \n U.S.C. 154(b) by 0 days. \n (21) Appl. No.: 14/607,201 \n (22) Filed: Jan. 28, 2015 \n (65) Prior Publication Data \n US 2015/O138332 A1 May 21, 2015 \n Related U.S. Application Data \n (60) Division of application No. 13/738,655, filed on Jan. \n 10, 2013, now Pat. No. 8,976,237, and a continuation of application No. 13/101,149, filed on May 5, 2011, \n now Pat. No. 8.373,753, and a division of application (*) Notice: \n (Continued) \n (51) Int. Cl. \n H04N 7/8 (2006.01) \n A6 IB I/00 (2006.01) \n (Continued) \n (52) U.S. Cl. \n CPC H04N 7/18 (2013.01); G06K 9/00 (2013.01); G06K 9/00288 (2013.01); \n (Continued) \n (58) Field of Classification Search \n USPC ........... 348/77; 340/506; 358/143, 147, 161, \n 358/169; 707/4, 103: 382/103, 209, 276, 382/277, 289, 291, 293, 294, 295, 282,305, \n 382/115, 107, 190 See application file for complete search history. \n (56) References Cited \n U.S. PATENT DOCUMENTS \n 5,258,837 A \n 5,473,369 A 11/1993 Gormley \n 12, 1995 Abe \n (Continued) \n FOREIGN PATENT DOCUMENTS \n WO 2007/044037 A1 4/2007 \n OTHER PUBLICATIONS \n PCT/US05/44656 International Search Report and Written Opinion, \n Jun. 26, 2006. \n (Continued) \n Primary Examiner — Jerome Grant, II \n (74) Attorney, Agent, or Firm — Dean D. Small: The Small \n Patent Law Group, LLC. \n (57) ABSTRACT \n A Surveillance system is provided that includes at least one sensor disposed in a security area of a Surveillance region to sense an occurrence of a potential security breach event; a plurality of cameras is disposed in the Surveillance region; at \n least one camera thereof has a view of the security area and can be configured to automatically gather biometric infor mation concerning at least one subject person in the vicinity of the security area in response to the sensing of a potential \n security breach event; one or more other of the plurality of cameras can be configured to search for the at least one Subject person; a processing system is programmed to \n produce a dossier corresponding to the at least one subject \n person to match biometric information of one or more persons captured by one or more of the other cameras with corresponding biometric information in the dossier. \n 35 Claims, 8 Drawing Sheets \n Apply Monitoring Rule on Security Area \n configure device and Security Area specific parameters (Door, \n restricted area, etc.) \n Environment \n (Map) Admin \n Security Area \n Admin display a sist of available Monitoring rules to apply to security area (Based on device characteristics Display a list of available devices to selecticeselect for applying Monitoring rule Select Security Area to configure Monitoring rule \n Monitoring rule (Security Function) \n Admin Configure Schedule & \n Eable Rule to start monitoring configured \n area Setect Cotstation Monitoring rule' metadata={'source': 'https://patentimages.storage.googleapis.com/98/f0/22/ae169ef67be168/US9432632.pdf', 'page': 0}


Die meisten Abstract scheinen mit "ABSTRACT \n" zu beginnen und endet beim nächsten "metadata="

In [17]:
print(type(pdf_list[1]))

<class 'langchain_core.documents.base.Document'>


Die enden nicht bei metadata, sondern die Class hängt die Metadaten am Ende einfach an. Deshalb "enden" die immer mit metadata.

In [18]:
abstract = str(test_py)
print(abstract.find("ABSTRACT"))
removeHeadline = abstract.find("ABSTRACT")

1898


In [19]:
print(abstract[removeHeadline:])

ABSTRACT \n A Surveillance system is provided that includes at least one sensor disposed in a security area of a Surveillance region to sense an occurrence of a potential security breach event; a plurality of cameras is disposed in the Surveillance region; at \n least one camera thereof has a view of the security area and can be configured to automatically gather biometric infor mation concerning at least one subject person in the vicinity of the security area in response to the sensing of a potential \n security breach event; one or more other of the plurality of cameras can be configured to search for the at least one Subject person; a processing system is programmed to \n produce a dossier corresponding to the at least one subject \n person to match biometric information of one or more persons captured by one or more of the other cameras with corresponding biometric information in the dossier. \n 35 Claims, 8 Drawing Sheets \n Apply Monitoring Rule on Security Area \n configure devi

In [20]:
abstractOhneHeadline = abstract[removeHeadline+12:] #+12 um Überschrift "ABSTRACT \n " ebenfalls zu entfernen
print(abstractOhneHeadline)

A Surveillance system is provided that includes at least one sensor disposed in a security area of a Surveillance region to sense an occurrence of a potential security breach event; a plurality of cameras is disposed in the Surveillance region; at \n least one camera thereof has a view of the security area and can be configured to automatically gather biometric infor mation concerning at least one subject person in the vicinity of the security area in response to the sensing of a potential \n security breach event; one or more other of the plurality of cameras can be configured to search for the at least one Subject person; a processing system is programmed to \n produce a dossier corresponding to the at least one subject \n person to match biometric information of one or more persons captured by one or more of the other cameras with corresponding biometric information in the dossier. \n 35 Claims, 8 Drawing Sheets \n Apply Monitoring Rule on Security Area \n configure device and Secur

In [21]:
split_abstr = abstractOhneHeadline.split('\\n ')
print(split_abstr)
abstractOhneN = "".join(split_abstr)
print(abstractOhneN)

['A Surveillance system is provided that includes at least one sensor disposed in a security area of a Surveillance region to sense an occurrence of a potential security breach event; a plurality of cameras is disposed in the Surveillance region; at ', 'least one camera thereof has a view of the security area and can be configured to automatically gather biometric infor mation concerning at least one subject person in the vicinity of the security area in response to the sensing of a potential ', 'security breach event; one or more other of the plurality of cameras can be configured to search for the at least one Subject person; a processing system is programmed to ', 'produce a dossier corresponding to the at least one subject ', 'person to match biometric information of one or more persons captured by one or more of the other cameras with corresponding biometric information in the dossier. ', '35 Claims, 8 Drawing Sheets ', 'Apply Monitoring Rule on Security Area ', 'configure device 

In [22]:
import re
finddigit = re.search(r"\d",abstractOhneN)
x = finddigit.start()
clean_abstract = abstractOhneN[:x]
print(clean_abstract)

A Surveillance system is provided that includes at least one sensor disposed in a security area of a Surveillance region to sense an occurrence of a potential security breach event; a plurality of cameras is disposed in the Surveillance region; at least one camera thereof has a view of the security area and can be configured to automatically gather biometric infor mation concerning at least one subject person in the vicinity of the security area in response to the sensing of a potential security breach event; one or more other of the plurality of cameras can be configured to search for the at least one Subject person; a processing system is programmed to produce a dossier corresponding to the at least one subject person to match biometric information of one or more persons captured by one or more of the other cameras with corresponding biometric information in the dossier. 


Hier folgt der die Abstract-Bereinigung nochmals, jedoch als Scheleife für alle Abstracts:

In [23]:
counter = 0
for i in pdf_list:
    counter += 1
    abstract = str(i)
    removeHeadline = abstract.find("ABSTRACT")
    abstractOhneHeadline = abstract[removeHeadline+12:] #+12 um Überschrift "ABSTRACT \n " ebenfalls zu entfernen
    #print(abstractOhneHeadline)
    split_abstr = abstractOhneHeadline.split('\\n ')
    abstractOhneN = "".join(split_abstr)
    finddigit = re.search(r"\d",abstractOhneN)
    x = finddigit.start()
    clean_abstract = abstractOhneN[:x]
    if (clean_abstract.count("") > 5):
        print("#",counter,"Clean Abstract:",clean_abstract)
    else:
        print("#",counter,"Failed reading PDF.")



# 1 Clean Abstract: A system is provided for localizing parts of an object in an image by training local detectors using labeled image exem plars with fiducial points corresponding to parts within the image. Each local detector generates a detector score corre sponding to the likelihood that a desired part is located at a given location within the image exemplar. A non-parametric global model of the locations of the fiducial points is gener ated for each of at least a portion of the image exemplars. An input image is analyzed using the trained local detectors, and a Bayesian objective function is derived for the input image from the non-parametric model and detector scores. The Bayesian objective function is optimized using a consensus of global models, and an output is generated with locations of the fiducial points labeled within the object in the image. 
# 2 Clean Abstract: A Surveillance system is provided that includes at least one sensor disposed in a security area of a Surveilla

<div class="alert alert-block alert-info">
<b>Ab hier folgt die Überprüfung der Ergebnisse; Lösung 1: Abgleich der Key Words und Classen</b> 
</div>

In [24]:
print(clean_abstract)

Mobile phones and other portable devices are equipped with a variety of technologies by which existing functionality can be improved , and new functionality can be provided . Some aspects relate to visual search capabilities , and determining appropriate actions responsive to different image inputs . Others relate to processing of image data . Still others concern metadata generation , processing , and representa tion . Yet others concern user interface improvements . Other aspects relate to imaging architectures , in which a mobile phone's image sensor is one in a chain of stages that successively act on packetized instructions / data , to capture and later process imagery . Still other aspects relate to distribution of processing tasks between the mobile device and remote resources ( “ the cloud ” ) . Elemental image pro cessing ( e.g. , simple filtering and edge detection ) can be performed on the mobile phone , while other operations can be referred out to remote service providers 

In [25]:
prompt1 = f"""```{clean_abstract}```\
The abstract above describes a concept for a novel invention.\
I would like to search a patent database to find out whether \
there are already patents for such a concept. Name 5 phrases that I can \
use for the search. Each phrase should contain between 5 to 10 words. \
Optimize the phrases to get back more results.
"""

prompt2 = f"""```{clean_abstract}```\
The abstract above describes a concept for a novel invention.\
I would like to search a patent database to find out whether \
there are already patents for such a concept. Please list me the codes of the 5 most relevant \
USPTO classifications to a possible patent for this concept without explanations for the codes.
"""

In [26]:
response_keywords_new = get_completion(prompt1)

In [27]:
response_classes_new = get_completion(prompt2)

In [28]:
print(response_keywords_new)
print("----")
print(response_classes_new)

1. "Visual search capabilities in mobile devices"
2. "Image data processing in portable devices"
3. "Metadata generation and representation in smartphones"
4. "User interface improvements in mobile phones"
5. "Imaging architectures in mobile phone's sensor"
----
1. H04W 4/00
2. G06F 3/01
3. G06K 9/00
4. G06T 7/00
5. H04N 5/232


In [29]:
prompt1 = f"""```{response_keywords}```\
Compare the list above with the following list. Just tell without explanation, how similar they are in percentage.\
```{response_keywords_new}```
"""

In [30]:
response_keywords_verification = get_completion(prompt1)

In [31]:
print("Key Word-Übereinstimmung:",response_keywords_verification)

Key Word-Übereinstimmung: 10%


In [32]:
prompt2 = f"""```{response_classes}```\
Compare the list above with the following list. Just tell me without explanation, how similar they are in percentage.\
```{response_classes_new}```
"""

response_classes_verification = get_completion(prompt2)

In [33]:
print("Klassenübereinstimmung:",response_classes_verification)

Klassenübereinstimmung: 20%


<div class="alert alert-block alert-info">
<b>Ab hier folgt die Überprüfung der Ergebnisse; Lösung 2: Abgleich der Abstracts insgesamt</b> 
</div>

In [37]:
our_abstract = pdf_abstract
print(pdf_list[1])

page_content='(12) United States Patent \n Cheng et al. USOO9432632B2 \n US 9.432,632 B2 \n Aug. 30, 2016 (10) Patent No.: \n (45) Date of Patent: \n (54) ADAPTIVE MULTI-MODAL INTEGRATED \n BOMETRIC IDENTIFICATION AND \n SURVELLANCE SYSTEMS \n (71) Applicant: Proximex Corporation, Cupertino, CA \n (US) \n (72) Inventors: Ken P. Cheng, Saratoga, CA (US); \n Edward Y. Chang, Santa Barbara, CA (US); Yuan-Fang Wang, Goleta, CA \n (US) \n (73) Assignee: Proximex Corporation, Cupertino, CA \n (US) \n Subject to any disclaimer, the term of this patent is extended or adjusted under 35 \n U.S.C. 154(b) by 0 days. \n (21) Appl. No.: 14/607,201 \n (22) Filed: Jan. 28, 2015 \n (65) Prior Publication Data \n US 2015/O138332 A1 May 21, 2015 \n Related U.S. Application Data \n (60) Division of application No. 13/738,655, filed on Jan. \n 10, 2013, now Pat. No. 8,976,237, and a continuation of application No. 13/101,149, filed on May 5, 2011, \n now Pat. No. 8.373,753, and a division of application (*

In [46]:
prompt1 = f"""```{our_abstract}```\
Compare the abstract from the text above with the following abstract. Just tell me without explanation, how similar they are in percentage.\
```{pdf_list[1]}```
"""

response_abstract_comparison = get_completion(prompt1)

In [48]:
print("Abstract Übereinstimmung:",response_abstract_comparison)

Abstract Übereinstimmung: 20%
